Josh preparations

flip1995 opened the pull request:
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust-clippy/pull/12759)*

### [Implement sync subcommand in clippy_dev](https://github.com/rust-lang/rust-clippy/pull/12759/commits/9abdb2743a60b875324aa5c7f6539e17fc8dd269)

Now that JOSH is used to sync, it is much easier to script the sync process.
This introduces the two commands `sync pull` and `sync push`. The first one will
pull changes from the Rust repo, the second one will push the changes to the
Rust repo. For details, see the documentation in the book.

### [Implement release subcommand in clippy_dev](https://github.com/rust-lang/rust-clippy/pull/12759/commits/4e497923873ff11ea88c402c1ae6323c7df55f52)

This is a QoL improvement for doing releases. The `release bump_version`
subcommand increments the version in all relevant `Cargo.toml` files. The
`release commit` command will only work with the new Josh syncs. Until then the
old way, using `git log` has to be used.

---

changelog: none

RalfJung commented on clippy_dev/src/sync.rs:
```diff
@@ -0,0 +1,252 @@
+}
+
+pub fn rustc_push(rustc_path: String, github_user: &str) {
+    const BRANCH: &str = "clippy-subtree-update";
```
At least for initial experimentation, it is quite useful to be able to overwrite the branch. May be worth to make this a flag with a default value.

RalfJung replied:
:shrug: up to you, for testing I sometimes want to have multiple branches to push to, to simulate "imagine we push to rustc, changes happen there, changes happen locally, we pull from rustc and then we push again".

flip1995 replied:
Makes sense. I'll add this then as a defaulted flag.

RalfJung commented on clippy_dev/src/sync.rs:
```diff
@@ -0,0 +1,230 @@
+}
+
+fn rustc_hash() -> String {
+    std::env::remove_var("RUSTUP_TOOLCHAIN");
```
You can avoid the unsafe operation by only removing the var in `sh`

flip1995 replied:
Done.

RalfJung commented on clippy_dev/src/sync.rs:
```diff
@@ -0,0 +1,230 @@
+
+fn rustc_hash() -> String {
+    let sh = Shell::new().expect("failed to create shell");
+    sh.set_var("RUSTUP_TOOLCHAIN", "");
```
```suggestion
    // Make sure we pick up the updated toolchain (usually rustup pins the toolchain
    // inside a single cargo/rustc invocation via this env var).
    sh.set_var("RUSTUP_TOOLCHAIN", "");
```

flip1995 replied:
Thanks! Now that everything is working, I will go through that code again and clean everything up.

flip1995 committed:
Implement sync subcommand in clippy_dev

Now that JOSH is used to sync, it is much easier to script the sync process.
This introduces the two commands `sync pull` and `sync push`. The first one will
pull changes from the Rust repo, the second one will push the changes to the
Rust repo. For details, see the documentation in the book.

flip1995 committed:
Implement release subcommand in clippy_dev

This is a QoL improvement for doing releases. The `release bump_version`
subcommand increments the version in all relevant `Cargo.toml` files. The
`release commit` command will only work with the new Josh syncs. Until then the
old way, using `git log` has to be used.

flip1995 commented:
https://github.com/flip1995/rust-clippy/pull/8

First sync has over 10k commit :open_mouth: See https://github.com/josh-project/josh/issues/1328

flip1995 commented:
Pro/Con list for Josh:

Pros:
- First sync takes ~1min instead of hours
- Later syncs take seconds, instead of minutes
- No more massive merge-commits, becoming normal commits, messing up `git blame` and git history.
- Scriptable (this PR), so sync can easily be done by anyone
- No more patched script from some git/git PR (kinda, see cons)

Cons:
- Creates a lot of commits: https://github.com/flip1995/rust-clippy/pull/8 https://github.com/josh-project/josh/issues/1328 (Those would be 1/3 of Clippy's history)
- Patch still required: https://github.com/josh-project/josh/pull/1326 (But maintainers said they want to work merging it and it can easily be installed through `cargo`, rather than `curl`ing the script and placing it in `/lib/git-core` or similar things)
- SHA1s in Rust repo no longer match SHA1s in Clippy repo. (The only time we needed that was for determining release commits, which can be worked around by including the sync commit in the Rust PR description. This is implemented in this PR)

The only big con is the number of commits. All other things, the pros outweigh the cons IMO.

RalfJung commented:
I guess it is trading off a few huge commits against a huge number of small merge commits. On the plus side, the josh merge commits do not pollute git blame.

RalfJung commented:
https://github.com/josh-project/josh/pull/1329 landed, so a josh patch shouldn't be required any more (but it'd be good to re-test this with the latest josh master to make sure that the final PR that landed still works as intended).

flip1995 commented:
Thanks, I will re-test this after the next sync on Thursday.

flip1995 commented:
From the meeting: The move to Josh got accepted. It would be nice to reduce the number of commits though. I will follow up on this and the move once I'm back from my business trip.

bors commented:
:umbrella: The latest upstream changes (presumably #12813) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #12867) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #13086) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

tgross35 commented:
It seems like you have been keeping this very up to date - is this just waiting on a final decision? Or is https://github.com/josh-project/josh/issues/1328 considered a blocker.

flip1995 commented:
We want to move to Josh, that's already decided. I didn't have the time to follow up with https://github.com/josh-project/josh/issues/1328. But IIUC if that should get implemented, we'd need to force push over the Clippy master, if we want to switch to it later.

RalfJung commented:
> But IIUC if that should get implemented, we'd need to force push over the Clippy master, if we want to switch to it later.

It's unclear, maybe there are ways to avoid that, e.g. using the `rev` filter.

bors commented:
:umbrella: The latest upstream changes (presumably #13157) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably 0712689aae115be49b7e429a346055e2849d1909) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

flip1995 committed:
New filter

RalfJung commented on clippy_dev/src/sync.rs:
```diff
@@ -1,8 +1,101 @@
-pub fn update_nightly() {
+use xshell::{Shell, cmd};
+
+const JOSH_FILTER: &str = r#":~(history="keep-trivial-merges,no-splice")[:rev(<=3382a954b74f79777805baf5b1c0a9a39a5e81fe:prefix=/home/pkrones/rust-lang/rust/src/tools/clippy,<=7836c8bb0589f20a694675b6efd5eb696f2de593:SQUASH)]://home/pkrones/rust-lang/rust/src/tools/clippy"#;
```
Something definitely went wrong here, your /home path should not appear.

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust-clippy/12759/3382a954b74f79777805baf5b1c0a9a39a5e81fe..60e98a816ed7d3b8d754def3046581a1b9b9588d)*

flip1995 replied:
Yup, I'll have to clean this up and test it once https://github.com/rust-lang/rust/pull/159789 is part of the nightly. Let's hope tomorrow.

flip1995 replied:
I think those paths appeared in the filter, because I used Christian's script with absolute paths :shrug:
